test/e2e/apimachinery/crd_validation_ratcheting.go

```diff
@@ -0,0 +1,898 @@
+			map:
+				foo:
+					field: "foo"
+				bar:
+					field: "foo"
+		`)
+		framework.ExpectNoError(err, "parsing test resource")
+		framework.ExpectNoError(applyPatch(testCRDGVR, "test-resource", instance.Object), "failed creating test resource")
+
+		ginkgo.By("updating CRD schema with constraints on correlatable fields to make instance invalid")
+		sch, err := parseSchema(`
+			type: object
+			properties:
+				field: 
+					type: string
+					maxLength: 5
+					x-kubernetes-validations:
+					- rule: self != oldSelf
+				struct:
+					type: object
+					properties:
+						field:
+							type: string
+							maxLength: 5
+							x-kubernetes-validations:
+							- rule: self != oldSelf
+				list:
+					type: array
+					maxItems: 5
+					x-kubernetes-list-type: map
+					x-kubernetes-list-map-keys: [key]
+					items:
+						type: object
+						properties:
+							key: {type: string}
+							field:
+								type: string
+								maxLength: 5
+								x-kubernetes-validations:
+								- rule: self != oldSelf
+						required:
+						- key
+				map:
+					type: object
+					maxProperties: 5
+					additionalProperties:
+						type: object
+						properties:
+							field:
+								type: string
+								maxLength: 5
+								x-kubernetes-validations:
+								- rule: self != oldSelf
+		`)
+
+		framework.ExpectNoError(err, "parsing schema")
+		framework.ExpectNoError(updateCRDSchema(testCRDGVR, *sch), "failed to update schema")
+
+		ginkgo.By("updating a label on the test resource")
+		instance.SetLabels(map[string]string{
+			"foo": "bar",
+		})
+		err = applyPatch(testCRDGVR, "test-resource", instance.Object)
+		for _, fieldPath := range []string{
+			"field",
+			"struct.field",
+			"list[0].field",
+			"list[1].field",
+			"map[foo].field",
+			"map[bar].field",
+		} {
+			gomega.Expect(err).To(gomega.MatchError(gomega.ContainSubstring(fieldPath)))
+		}
+	})
+
+	// ginkgo.It("MUST evaluate a CRD Validation Rule with oldSelf = nil for new values when optionalOldSelf is true", func() {
+	// })
+
+})
+
+func parseSchema(source string) (*apiextensionsv1.JSONSchemaProps, error) {
+	source, err := fixTabs(source)
+	if err != nil {
+		return nil, err
+	}
+
+	d := utilyaml.NewYAMLOrJSONDecoder(strings.NewReader(source), 4096)
+	props := &apiextensionsv1.JSONSchemaProps{}
+	return props, d.Decode(props)
+}
+
+func parseUnstructured(source string) (*unstructured.Unstructured, error) {
+	source, err := fixTabs(source)
+	if err != nil {
+		return nil, err
+	}
+
+	d := utilyaml.NewYAMLOrJSONDecoder(strings.NewReader(source), 4096)
+	obj := &unstructured.Unstructured{}
+	return obj, d.Decode(&obj.Object)
+}
+
+// fixTabs counts the number of tab characters preceding the first
+// line in the given yaml object. It removes that many tabs from every
+// line. It returns error (it's a test function) if some line has fewer tabs
+// than the first line.
+//
+// The purpose of this is to make it easier to read tests.
+func fixTabs(in string) (string, error) {
```

Comment:
Should we open a help-wanted issue to introduce a `stripMargin` style function (Scala has a good one) somewhere common to the larger codebase? I really like how this worked out here.

Reply:
Yeah, we also use something similar in `kube-openapi`, and `kubectl-validate` has a similar need....I wonder if k8s.io/utils would be a good place?